File: algorithms/1.TwoSum.py
"""
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target. # noqa

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

https://leetcode.com/problems/two-sum/description/
"""


class Solution:
    def twoSum(self, nums: list, target: int) -> list:
        # [2,7,11,15] / [3,2,3]

        # Crear un hash map para almacenar los valores y los indices de la lista # noqa
        hash_map = {}

        # se recorre toda la lista
        for i, num in enumerate(nums):
            # sacamos la diferencia del target vs cada valor de la lista
            diff = target - num

            # validamos que este en nuestro hash map
            if diff in hash_map:
                return [hash_map[diff], i]

            # si no esta, se agrega
            hash_map[num] = i

        return None


numbers = [1, 2, 3, 4, 5, 6]
target = 6
sol = Solution()
print(sol.twoSum(nums=numbers, target=target))

File: leetcode/algorithms/1.TwoSum.cpp
/*
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 
Example 1:
                 k
Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

Example 2:
Input: nums = [3,2,4], target = 6
Output: [1,2]

Example 3:
Input: nums = [3,3], target = 6
Output: [0,1]
*/

#include <iostream>
#include <vector>

void printVector(std::vector<int>nums);

class Solution {
public:
    std::vector<int> twoSumBruteForce(std::vector<int>& nums, int target) {
        std::vector<int> result;
        int tmp;
        for(int i=0; i < nums.size(); i++){
            tmp = target - nums[i]; // 7
            for(int k=0; k<nums.size(); k++){
                if(nums[k] == tmp && i != k){
                    result.push_back(i);
                    result.push_back(k);
                    return result;
                }
            }            
        }
        return result;
    }

      /*            i
        nums   = [2,7,11,15], target = 9
        map    = {2: 0}
        result = [0,1] ✅
    */
    std::vector<int> twoSum(std::vector<int>& nums, int target){
        int tmp;
        std::vector<int> result;
        std::unordered_map<int, int> map;

        for(int i=0; i < nums.size(); i++){
            tmp = target - nums[i]; // 2
            if(map.find(tmp) != map.end()){
                result.push_back(map[tmp]);
                result.push_back(i);
                return result;
            }else{
                map[nums[i]] = i;
            }
        }
        return result;
    }
};


int main(){
    Solution sol;
    std::vector<int> nums;
    int target;

    nums = {2,7,11,15};
    target = 9;
    printVector(sol.twoSum(nums, target));

    nums = {3,2,4};
    target = 6;
    printVector(sol.twoSum(nums, target));

    nums = {3,3};
    target = 6;
    printVector(sol.twoSum(nums, target));

    return 0;
}


void printVector(std::vector<int>nums){
    for(int i = 0; i < nums.size(); i++){
        std::cout<< nums[i] << ", ";
    }
    std::cout<<"\n";
}
